[Core Data] Make new Core Data migration path. Remove models below 30

iamgabrielma opened the pull request:
Apologies for the 15k LOC, it belongs mostly to deleting the `.xcdatamodels` and associated `.xcmappings`, the actual changes are the ~200 additions.

## Description
The app currently has 124+ Core Data model versions, which represents 7+ years of incremental migrations. This PR starts the work on clearing the app from core data old models as part of p91TBi-drE-p2 by triggering a new migration path when a merchant hits the threshold (model 30). In future PRs we'll be raising the threshold to get it closer to the latest version.

- If a merchant is below the threshold (<30), the new path is triggered: The database is destroyed, a fresh one set to the latest model is created, and the merchant is migrated directly, instead of performing iterative migrations.
- If a merchant is above the threshold (30+), the existing path of iterative migration is triggered.

I left a bunch of DDLogs with what's happening, following the style of other core data handling functions, and through the migration steps.

## Changes
- Created a `CoreDataMigratorUtils` to encapsulate the logic related to finding the current DB version
- Updated `CoreDataIterativeMigrator` with the new migration path. This relies on the existing persistent store coordinator to automatically create a fresh store with the latest model as needed.
- Deleted oldest 30 `.xcdatamodel `
- Deleted associated `.xcmappings` migrations between several of these models
- Updated existing tests, added new tests, and updated documentation.

## Steps to reproduce/Testing information
This is not very straight-forward to test, ultimately we'd want the merchant to not crash if they happen to be on a model version that has been deleted when a migration is performed. This crash is expected with an iterative migration if a step is missing, but won't happen with the direct migration. 

If the worst happens and they crash due being unable to perform iterative migration, the app already performs a direct migration as a fallback, which is what we leverage directly via the threshold check when we can.

I added instructions on the P2 at p91TBi-drE-p2 on how I initially tested it by deleting 120 models instead, since this is the oldest available in TestFlight it allows us to simulate what would happen when a migration is needed and the threshold is hit. 

At the moment of writing this PR the latest available TF version is 22.3, which as per MIGRATIONS.md was using model 120, so I'll use that as example. You can:
- Install oldest TF version, login and play around, this should set the model 120 in your device
- Set `oldestSupportedDataModel` to 122 or higher (including the model version that your TF build should be in), so it determines that anything below is an old model and should be purged entirely.
- Delete the remaining `.xcdatamodels` up till 122 in Xcode, then run the app from Xcode in the same device you previously ran the TF build. 
- Following the traditional path, we would see a crash here since the iterative migration is broken (missing steps). Instead, since the model is below the threshold, we should migrate directly to the latest version, skipping the iterative migration path.
- Observe there is no crash when running the app, and that works normally. From testing, the Order list will appear empty, a simple pull-to-refresh should repopulate it.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
Create CoreDataMigratorUtils and tests

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
check for destroyPersistentStore. Extract utils to own file

iamgabrielma committed:
lint

iamgabrielma committed:
update oldest supported DB threshold to 30

this will make some test fail (expected), ie: test_iterativeMigrate_can_iteratively_migrate_from_model_1_to_model_10 since we cannot iterateively migrate anymore below 30

iamgabrielma committed:
remove models 1 to 29

iamgabrielma committed:
updat ManagedObjectModelsInventoryTests

iamgabrielma committed:
remove mapping models below 30 and related migration tests

iamgabrielma committed:
lint

iamgabrielma committed:
update docs

iamgabrielma committed:
update tests that rely on state from removed models

iamgabrielma commented on Modules/Sources/Storage/CoreData/CoreDataMigratorUtils.swift:
```diff
@@ -0,0 +1,38 @@
+
+        // Fallback: couldn't parse version, assume it's very old (version 0)
+        DDLogWarn("[CoreDataMigratorUtils] Could not parse version number from '\(versionName)'. Assuming version 0.")
+        return 0
```
I left this to assume 0 so it does not depend on the implementation details of which model is the latest. Zero will always be less than the threshold, so it will always trigger an iterative migration if we fail to name a model properly, which should never happen.

iamgabrielma commented on Modules/Sources/Storage/Model/MIGRATIONS.md:
```diff
@@ -459,50 +459,50 @@ This file documents changes in the WCiOS Storage data model. Please explain any
 - Delete `OrderStatsItem` entity
 
-## Model 29 (Release 4.7.0.0)
+## Model 29 (Release 4.7.0.0) - Deleted
```
Maybe we can delete these too, for now I just added a "deleted" hint.

jaclync replied:
Keeping them with a deleted note feels better so that developers know why it doesn't start with version 30.

iamgabrielma commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift:
```diff
@@ -187,12 +187,12 @@ final class CoreDataIterativeMigratorTests: XCTestCase {
+        // 30 → 31 → ... → 40
         //
-        XCTAssertEqual(spyCoordinator.destroyedURLs.count, 9)
+        XCTAssertEqual(spyCoordinator.destroyedURLs.count, 10)
```
There was no Model 1. It went from Model to Model 2 hence why I changed the number of steps to assert to 10

jaclync replied:
I believe "Model" was the equivalent of version 1. The count increase makes sense because the previous 1-10 was 10 versions, and 30-40 is 11 versions.

iamgabrielma commented:
Only one review is needed. No rush to review this before 23.0 freeze.

jaclync commented:
Apology for not getting to it today. I will take a look first thing next Monday if it's still open.

jaclync commented on Modules/Sources/Storage/CoreData/CoreDataIterativeMigrator.swift:
```diff
@@ -176,4 +187,37 @@ private extension CoreDataIterativeMigrator {
+        } catch {
+            DDLogError("[CoreDataIterativeMigrator] Failed direct migration. Will fall back to iterative migration. Error \(error)")
+        }
+    }
```
nit: the logging here feels duplicate to the [call site](https://github.com/woocommerce/woocommerce-ios/pull/15987/files#diff-2b281e7b3f3bb3efc499c7ee61240e9bb35334a426fff2b945d1464efd18a8f6R66-R70). Could the call site call exactly the same code here as the destroy function is just one line?

iamgabrielma replied:
Yeah, the function doesn't add much value at this point since it's just wrapping the next step. Simplified on https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/d1d463b8ff01630765ca330ec66a2a30c30955eb

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigrator+MigrationStepTests.swift:
```diff
@@ -24,8 +24,8 @@ final class CoreDataIterativeMigrator_MigrationStepTests: XCTestCase {
-        let modelVersion23 = ModelVersion(name: "Model 23")
-        let modelVersion31 = ModelVersion(name: "Model 31")
+        let modelVersion23 = ModelVersion(name: "Model 33")
+        let modelVersion31 = ModelVersion(name: "Model 41")
```
❓ I'm a bit confused by why these model versions need to be incremented by 10, instead of removing the ones < 30?

nit: for all the diffs in this file, the constant names can be updated to match the model version name

iamgabrielma replied:
> I'm a bit confused by why these model versions need to be incremented by 10, instead of removing the ones < 30?

No specific reason, I just added +10 to all of the cases for consistency. Since model 23 is below the current threshold it will return nil and not trigger the iterative migration for the test. Or are you referring to something else? 

> the constant names can be updated to match the model version name

Uggh of course I forgot, thanks! 😅 Updated https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/66da437e659618de7646382fbaad33c2f8a0a0de

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigrator+MigrationStepTests.swift:
```diff
@@ -131,22 +131,4 @@ final class CoreDataIterativeMigrator_MigrationStepTests: XCTestCase {
-    /// This seems like a bug in the `steps()` loop that has existed for a long time. I would have
-    /// expected that 0 steps are returned. I'm just keeping it as is for now. We don't
-    /// reach this condition because of the precondition checks in `CoreDataIterativeMigrator`.
-    func test_steps_returns_source_to_latest_version_MigrationSteps_if_the_source_and_target_are_the_same() throws {
```
❓ Any reasons for removing this test case? This seems like a valid test case to test the behavior when migrating to the same model version.

iamgabrielma replied:
My error here, it is a good test that I removed temporarily, as hardcoding the model to be asserted no longer worked when we removed the initial batch, so I needed to rewrite it but never did. 

Restored and updated here: https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/dc6431f972dd14359b0476882d637d589f666935

jaclync commented on Modules/Sources/Storage/CoreData/CoreDataIterativeMigrator.swift:
```diff
@@ -176,4 +187,37 @@ private extension CoreDataIterativeMigrator {
+        }
+        // If the merchant is on a model version older than this, nuke the DB and migrate them to the latest one without iterative process
+        let oldestModelThreshold = Constants.oldestSupportedDataModel
+        let versionNumber = CoreDataMigratorUtils.extractVersionNumber(from: sourceVersion.name)
```
nit: how about combining `CoreDataMigratorUtils.findSourceVersion` and `CoreDataMigratorUtils.extractVersionNumber` into one function to simplify the code a bit? It looks like the two methods are only called here.

```suggestion
        guard let versionNumber = CoreDataMigratorUtils.findSourceVersionNumber(for: sourceModel, in: modelsInventory) else {
            DDLogInfo("Could not determine source model version. Using iterative migration.")
            return false
        }
        // If the merchant is on a model version older than this, nuke the DB and migrate them to the latest one without iterative process
        let oldestModelThreshold = Constants.oldestSupportedDataModel
```

iamgabrielma replied:
In general I'd prefer to keep these separated, but with the simplification you suggested on https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/608096441ab9aa77e9f5a0e71952487e9ff0318d we can also remove `extractVersionNumber` entirely.

jaclync commented on Modules/Sources/Storage/CoreData/CoreDataMigratorUtils.swift:
```diff
@@ -0,0 +1,38 @@
+    static func findSourceVersion(for sourceModel: NSManagedObjectModel,
+                                  in modelsInventory: ManagedObjectModelsInventory) -> ManagedObjectModelsInventory.ModelVersion? {
+        do {
+            let allModels = try modelsInventory.models(for: modelsInventory.versions)
+            for (index, model) in allModels.enumerated() {
+                if model.isEqual(sourceModel) {
+                    return modelsInventory.versions[index]
+                }
+            }
```
❓ Trying to ensure that it can return a deleted version. When I printed out `modelsInventory.versions` when launching the app in `CoreDataManager.migrateDataModelIfNecessary`, the versions are 30+; and `modelsInventory.models(for:)` maps the versions. Can a deleted version be returned, and thus be matched to a source model?

If it's not possible, I think the code can be simplified by not having to compare with the oldest model version constant, since the available models in `modelsInventory.versions` already imply the oldest version. If it makes sense, `shouldDestroyPersistentStore` can return true when a model version cannot be found, when this function returns nil.

iamgabrielma replied:
That's a very good catch, and excellent suggestion. We can just rely on the existence of the model rather than having to check the version, which simplifies the logic and allows us to remove some. Updated here: https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/608096441ab9aa77e9f5a0e71952487e9ff0318d

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift:
```diff
@@ -139,10 +139,10 @@ final class CoreDataIterativeMigratorTests: XCTestCase {
+        let model10 = try managedObjectModel(for: "Model 40")
 
-        let storeURL = try urlForStore(withName: "Woo Test 10.sqlite", deleteIfExists: true)
+        let storeURL = try urlForStore(withName: "Woo Test 30.sqlite", deleteIfExists: true)
```
❓ the deleted diff was using the model version of `model10`, which is now "40." I was expecting this `storeURL` to correspond to the later version, `"Woo Test 40.sqlite"`? A bit surprised this test case passes 🤔 

Also a nit to rename the `model*` constants now that the versions have changed.

iamgabrielma replied:
> the deleted diff was using the model version of model10, which is now "40." I was expecting this storeURL to correspond to the later version, "Woo Test 40.sqlite"? A bit surprised this test case passes

I think this is expected though: My understanding is that "Woo Test 30.sqlite" reflects that deliberately creates a persistent store with Model 30, saves it to a specific SQLite file, asserts that the DB exists (not nil), then attempts to open that same file with Model 40 on _"Load using a different model"_ which should fail (returns nil)

> Also a nit to rename the model* constants now that the versions have changed.

Oops! https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/08de9bf37c275b84e97c65dd182183fa932adca0

jaclync replied:
It looks like the URL name doesn't matter, I changed the name to a different model version and also remove the model version and the test passed.

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift:
```diff
@@ -272,6 +272,76 @@ final class CoreDataIterativeMigratorTests: XCTestCase {
         assertEmpty(spyCoordinator.destroyedURLs)
     }
+
+    func test_extractVersionNumber_when_no_version_number_then_handles_base_model() {
```
nit: maybe `CoreDataMigratorUtils` test cases can be moved to a separate test class (plus: Swift Testing, so that `test_extractVersionNumber__when_version_number_then_handles_numbered_models` can use arguments for the different combinations).

iamgabrielma replied:
Thanks for the suggestion! I removed the test along https://github.com/woocommerce/woocommerce-ios/pull/15987/commits/608096441ab9aa77e9f5a0e71952487e9ff0318d , but it's something I'll keep in mind for `findSourceVersion`

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataIterativeMigratorTests.swift:
```diff
@@ -272,6 +272,76 @@ final class CoreDataIterativeMigratorTests: XCTestCase {
+        XCTAssertEqual(versionNumber, 0)
+    }
+
+    func test_extractVersionNumber__when_version_number_then_handles_numbered_models() {
```
super nit:
```suggestion
    func test_extractVersionNumber_when_version_number_then_handles_numbered_models() {
```

iamgabrielma replied:
Removed with https://github.com/woocommerce/woocommerce-ios/commit/608096441ab9aa77e9f5a0e71952487e9ff0318d

jaclync commented on Modules/Tests/StorageTests/CoreData/CoreDataManagerTests.swift:
```diff
@@ -211,8 +211,9 @@ final class CoreDataManagerTests: XCTestCase {
-        XCTAssertNil(NSEntityDescription.entity(forEntityName: Note.entityName,
+        // The OrderFeeLine entity does not exist in Model 30, was introduced in Model 42
+        // This proves that the store was reset to Model 30.
+        XCTAssertNil(NSEntityDescription.entity(forEntityName: OrderFeeLine.entityName,
```
nice replacement 👍

jaclync reviewed (commented):
Thanks for taking on this task to reduce the app size. I had some questions marked with ❓ , especially the one about whether it can detect a deleted model version. Let me know what you think.

iamgabrielma committed:
Merge branch 'trunk' into task/WOOMOB-1006-remove-cd-models-1-to-30

iamgabrielma committed:
remove duplication due to wrapper

iamgabrielma committed:
update var names to match model name

iamgabrielma committed:
Restore test case when when migrating to the same model version

iamgabrielma committed:
simplify shouldDestroyPersistentStore check

Rather than relying on a set constant and comparing model versions, we can rely on the existence of the model itself in the build to simplify the logic

iamgabrielma committed:
update model names to fit their number

iamgabrielma committed:
lint

jaclync commented on Modules/Sources/Storage/CoreData/CoreDataIterativeMigrator.swift:
```diff
@@ -59,6 +59,18 @@ final class CoreDataIterativeMigrator {
+                try persistentStoreCoordinator.destroyPersistentStore(at: sourceStoreURL, ofType: storeType, options: nil)
+                DDLogInfo("[CoreDataIterativeMigrator] Database at \(sourceStoreURL) destroyed successfully.")
+            } catch {
+                DDLogError("[CoreDataIterativeMigrator] Direct migration failed. Error: \(error). Falling back to iterative migration.")
```
super nit: maybe the error message can match the operation that failed?
```suggestion
                DDLogError("[CoreDataIterativeMigrator] Database destruction failed. Error: \(error). Falling back to iterative migration.")
```

jaclync reviewed (approved):
Thanks for the updates, LGTM :shipit: Tested with TestFlight 22.3 build then PR build with model versions up to 121 deleted. Didn't see the DDLogInfo messages, but the app didn't crash and recovered after a bit 👍

iamgabrielma committed:
Update Modules/Sources/Storage/CoreData/CoreDataIterativeMigrator.swift

Co-authored-by: Jaclyn Chen <watertranquil@gmail.com>

iamgabrielma committed:
Merge branch 'trunk' into task/WOOMOB-1006-remove-cd-models-1-to-30
